ADR036: Arbitrary signature

fdymylja opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description
ADR for arbitrary message signature to allow cosmos-sdk users to create and verify messags, where by message we intend arbitrary bytes (json files, text, etc).

Tentative implementation: https://github.com/cosmos/cosmos-sdk/pull/7896

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

alessio commented:
CC'ing @antoineherzog

ethanfrey commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,60 @@
+
+The aim is being able to sign arbitrary messages, even using Ledger or similar HSM devices.
+
+As a result signed messages should look roughly like Cosmos SDK messages. chain-id, account_number and sequence can all be assigned invalid values.
```
:+1:

aaronc commented:
Is this just for cli users? Or would nodes do this verification too?

ethanfrey commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,60 @@
+### Negative
+
+- Probably this feature should be provided via the keystore 
+- Developing it as a module means we need to rely on using sign transactions in offline mode, and implementing it as default for the command means that we need to do a flag override (offline, account number, sequence, chain id)
```
Huh? This is a module? I thought this was just a format specification.

Can't you just define a format as above, and add a protobuf spec fo MsgSendText (and legacy amino compatibility for it)? The chain never needs to process this, as existing decorators would be sure to reject it.

amaurymartiny replied:
Agreed, it could go in x/auth.

ethanfrey reviewed (approved):
Interesting, so the signers / clients wouldn't have to change at all - this is just a standardized spec for how to make an invalid tx that serves an an off-chain signed message. I like it.

Only question is why this is a module and not just a proto definition.

ethanfrey commented:
I think the main users would not be cli tools, but rather keplr and other web-based signers. As one of the most common uses of signed messages in ethereum is to authenticate yourself to a webapp. (Please correct me if I am wrong, and ideally expand on this on context - give 1 or 2 concrete workflows that would be enabled)

antoineherzog reviewed (approved):
Great to see a signed message spec coming into the cosmos world 👍 
I agree with Ethan, I believe this spec is only about the format specification. The spec maybe could provide an example of a signed message. Up to the client to implement this spec. Ideally we want to have the major clients of the cosmos ecosystem like KeplR and other web-based clients to implement it. 

Signed message are usefull to login into a Cosmos Dapp. They are also usefull for external provider like KYB providers which wants to create certificates in the cosmos ecosystem with a cosmos key for corporate use cases. CLI client can be still be interesting for theses kind of providers who wants to be able to sign a message from a command line interface.

moshemalawach commented:
Implemented in :
- https://github.com/aleph-im/pyaleph
- https://github.com/aleph-im/aleph-js
- https://github.com/aleph-im/aleph-client

Example use with a cosmos account:
https://aleph-im.github.io/aleph-js/guide/getting-started.html#cosmos-cosmos-sdk

aaronc commented:
Why do we need a new Msg type? Could we just overload memo? Do we even need to make this a TC?

zmanian commented:
@aaronc this needs to look sufficiently like a transaction to be signable on a ledger or other existing custody solution.

Overloading memo is bad because we want to sign structured data.

alexanderbez commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,60 @@
+
+- 28/10/2020 - Initial draft
+
+## Authors
```
This should reference some original work you can find [here](https://github.com/cosmos/ics/pull/33).

fdymylja replied:
addressed

alexanderbez commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,60 @@
+- Frojdi Dymylja (@fdymylja)
+
+## Status
+Draft
```
```suggestion

Draft
```

fdymylja replied:
addressed

alexanderbez commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,60 @@
+
+Currently, in the SDK, there is no convention to sign arbitrary message like on Ethereum. We propose with this specification a way to sign arbitrary message for Cosmos SDK chains.
+
+## Context
```
```suggestion

## Context
```

fdymylja replied:
addressed

alexanderbez reviewed (changes requested):
I feel like a lot is missing from this. What does the structure look like? How is it encoded? How will an HSM sign it and display it? What kinda of formats/characters are acceptable?

fdymylja commented:
I've addressed the changes highlighted. Added the Msg definition and an example. 

> How is it encoded? 
If by encoded you mean serialized, then shouldn't we be able to support both JSON and Protobuf?

> How will an HSM sign it and display it?
Since this is an sdk.Msg, shouldn't what we have at our disposal work already?

liamsi commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+
+## Consequences
+
+There is a specification on how messages, that are not meant to be broadcast to an active 
```
This seems like the sentence isn't finished (at least a work is missing)?

fdymylja replied:
addressed, thank you for pointing this out

alessio commented:
@alexanderbez mind reviewing this again please?

aaronc commented:
How would this relate with #7491 ? Are we verifying the pubkey in the `Tx` or in state?

alexanderbez commented on docs/architecture/adr-036-arbitrary-signature.md:
````diff
@@ -0,0 +1,103 @@
+Proto definition:
+```proto
+// MsgSignedMessage defines 
+message MsgSignedMessage {
````
I believe this should be `MsgSignText` based on the previous text.

robert-zaremba replied:
I  would prefer `MsgSignData`. Msg is overloaded here (3 times in 2 "words")

alexanderbez commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+    // Signer is the sdk.AccAddress of the message signer
+    bytes Signer = 1 [(gogoproto.jsontag) = "signer", (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
+    // Message represents the raw bytes of the content that is signed (text, json, etc)
+    bytes Message = 2 [(gogoproto.jsontag) = "message"];
```
How will this be displayed on Ledgers and similar HSMs?

alessio replied:
I think at this we don't have an answer to that. Would you be OK with unblocking this on the promise that we'll continue doing work once this is approved?

robert-zaremba replied:
The we should add a further discussion secton in the ADR and create a tracking issue on Github

alessio replied:
> Then we should add a further discussion section in the ADR and create a tracking issue on Github

How could this reply of yours possibly answer my question?

marbar3778 replied:
Open an issue to facilitate further discussions. Let's get this merged. I would say the implementation is blocked until the issue on ledger and hsm is undersood and a solution has been agreed on. We can see if zondax can join a sdk call in the coming weeks.

alessio replied:
OK, thanks @marbar3778 and @robert-zaremba

ethanfrey commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+    bytes Signer = 1 [(gogoproto.jsontag) = "signer", (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
+    // Message represents the raw bytes of the content that is signed (text, json, etc)
+    bytes Message = 2 [(gogoproto.jsontag) = "message"];
+}
```
I would propose the following, based on both the OAuth2 [implicit](https://tools.ietf.org/html/rfc6749#section-4.2.1) and [code grant flows](https://tools.ietf.org/html/rfc6749#section-4.1.1).

```proto
message MsgSignText {
    // we can add other request_types later with different semantics - this is for the obvious "prove your identity" / access control flow
    string request_type = 1;  // MUST be "bauth"
   
    // which client we are identifying with, could be `registry.atlas.io` or `2du6Sje7` but is unique to the site you authenticate with 
    string client_id = 2;     

    // (OPTIONAL) app-dependent if we want to give a partial grant - eg. for a forum, we could auth it to upvote but not make posts
    string scope = 3;

   // (OPTIONAL) a nonce or session token that can prevent replay attacks
   string state = 4;

    // ????? I don't think we need this, but it is in the Oauth2 flow - I belive this is different in the proposed flow.
    // but maybe is useful, I would like an end-to-end client workflow explained.
    string redirect_url = 5;
}
```

Note that the `Signer` would only be needed for actual processing in a module, via `GetSigners()`, but since we don't ever intend this to go to the blockchain, I would remove it as unnecessary data. The pubkey stored in the signature is the information we want to prove. 

This would allow a website to produce an unsigned transaction *without knowing your pubkey* and then you can sign it with any pubkey you control to identify yourself.

ethanfrey commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+- Antoine Herzog (@antoineherzog)
+- Zaki Manian (@zmanian)
+- Aleksandr Bezobchuk (alexanderbez) [1]
+- Frojdi Dymylja (@fdymylja)
```
Can you please add me to the Authors list if you accept my new format and justification?

`- Ethan Frey (@ethanfrey)`

ethanfrey reviewed (commented):
I agree with Bez. The content of the message is too vague. Signer plus "some unspecified binary data". However, the Context here is also super vague "Having the ability to sign messages off-chain has proven to be a fundamental aspect of nearly any blockchain." doesn't state which messages for which purposes.

However, one main use case alluded to below is authentication using the same pubkey used on a blockchain. In such a case, I think it is good to model upon the learnings of OAuth2 (as that is more or less what you are trying to emulate and has had lots of security research). I provide a concrete example below, based on both [Authorization Code Grant](https://tools.ietf.org/html/rfc6749#section-4.1) and [Implicit Flow](https://tools.ietf.org/html/rfc6749#section-4.2) which are designed for JS clients.

If you wish to define another concrete use case beyond authorization (eg. state channel messages), I would define another message type there, or possibily extend this type with a different `request_type`  for that use-case.

The following should be sufficient to provide replay protection (both `client_id` - like `chain_id` - and `state` - like `sequence`). As well as to allow partial permisisoning on that client via `scope` rather than just an all-or-none permission.

Remember we are not only proving our identity to a dApp, but also granting a untrusted dApp a token that possibly allows it to act on our behalf with some web backend (in many use cases).

ethanfrey commented:
I don't really understand the [aleph.im example above](https://explorer.aleph.im/address/CSDK/star169z5k2fp957z25ynegp3tj4jdan74vgjply0k5/message/AGGREGATE/668c1430639b1e365caa7296be6cf3e72a9f3cc014ca88fb064483e5f6cdc006) but maybe you do not want id, but to sign each "POST".

In such case the above design may not make sense, but I would still keep minimally `client_id` and possibly `state` for replay protection.

```proto
message MsgSignText {
    // define new flow for signing one post
    string request_type = 1;  // MUST be "message"
   
    // which client we are identifying with, could be `registry.atlas.io` or `2du6Sje7` but is unique to the site you authenticate with 
    string client_id = 2;     

   // (OPTIONAL) a nonce or session token that can prevent replay attacks
   string state = 4;

    // this is a human-readable JSON-encoded message that will be posted to the client side using your identity
    string message = 6;
}
```

You could make a different Msg type for each case, or make one common message type, that always has `request_type`, `client_id` and `scope` and then has optional fields based on the `request_type`

fdymylja commented:
I think the initial intention of the ADR was to offer a specification to sign arbitrary data, let it be structured or as simple as a string, and make it look as similar as possible to a cosmos-sdk Msg type. 

The reasons for which it has to be a Msg type, from the discussion I've had with @antoineherzog, were the following:

- Make it signable via ledger (I'm not familiar with ledger , so I assume we're not capable of signing raw bytes, but we have to pass a cosmos sdk msg to sign?)

- Cosmos-sdk libraries can already sign and verify this message, without the need to implement anything new.

> However, one main use case alluded to below is authentication using the same pubkey used on a blockchain. In such a case, I think it is good to model upon the learnings of OAuth2

Whilst this is probably the main use case, and probably deserves an ADR on its own for this exact purpose, I think the aim of the current ADR is to offer something more abstract and primitive.  

For instance the OAuth2 Design could be encoded as raw bytes and saved into the Message byte field of the proposed specification, and the implementers are free to decode it and do their checks as they wish without being limited.

Then we could argue that a 'vague' signable bytes specification is not needed, in this case I'd let @antoineherzog and @zmanian add more concrete use cases which differ from each other and justify having a 'vague' specification.

alessio commented:
(CC'ing the Ledger Expert© @jleni)

ethanfrey commented:
> For instance the OAuth2 Design could be encoded as raw bytes and saved into the Message byte field of the proposed specification, and the implementers are free to decode it and do their checks as they wish without being limited.

The problem with this is that when signing, we need to present to the user a human-readable description of *what* they are signing. If it is some arbitrary encoded bytes, how can they give informed consent?

You can define it as "arbitrary UTF-8 encoded JSON" and assume they will be able to make heads or tails of that (in which case a string not `bytes`). Maybe that makes sense in a UI, and does give flexibility, but the UX would be sub-par.

I agree the OAuth case is too specific if that is not your desire. I don't really know your use case and my main point is something that can be displayed in a human-readable form.

zmanian commented:
IMO the scope of this should be limited to SIGN_MODE_LEGACY_AMINO and SIGN_MODE_DIRECT. 

UX improvements should be added with the future SIGN_MODE intended to replace SIGN_MODE_LEGACY_AMINO.

amaurymartiny commented:
> For instance the OAuth2 Design could be encoded as raw bytes and saved into the Message byte field of the proposed specification, and the implementers are free to decode it and do their checks as they wish without being limited.

I would advise against that. One big advantage of `Msg` being of type `Any` is that module developers can create well-structured `Msg`s for exactly their specific use case, to avoid what you described.

I think what's missing in my understanding of this ADR is: what are the use cases? What content do you expect to put generally in the `MsgSignedMessage`? If those use cases are common enough, we could create separate ADRs for them.

If human-readability is the issue, like Zaki said I think we should wait for SIGN_MODE_TEXTUAL, where human-readability comes for free for all `Msgs`s.

antoineherzog commented:
I believe use-cases are multiple like:

- Login into a cosmos dapps
- Create a certificate
- Payment Channel message
- Use layer-2 networks like aleph.im (or zk roll ups too)
- Prove ownership of your address to do a reverse lookup  

Ethereum opens way with signed message long time ago(https://web3js.readthedocs.io/en/v1.2.0/web3-eth-accounts.html#sign) and Dapps used it in many ways:  login to a dapps, dapps paying fees on behalf of their users, etc...

In our case, our blockchain Starname is using already the signed message for two use-cases:
- to store metadata on the Aleph Network
- to provide a command line for our KYB partner to be able to sign certificate

ethanfrey commented:
I think referencing 4 year old Ethereum APIs is great for inspiration, and I have used a lot of inspiration from erc20, erc777, erc721 and more when writing CosmWasm contracts. There are many good ideas there that have worked in practice.

However, many of these have been adopted and extended over time and we can learn from the various "best practices" around such APIs (open zeppelin usually has good material) and design a better API that enforces such best practices. (In the case of cw20, we use increment/decrement allowance rather than a new value as there was a known race condition with the erc20 api)

alessio commented:
Thanks @zmanian. Can we all agree to this as the use case that this ADR must focus on?

>I would like to be able to login to application while proving ownership of a cosmos address.
> I'd like to have a payload with a cosmos address, a url that I'm trying to login into, application specific key value metadata.
> I'd like to be able to support features like multisig addresses and I'd like to be able to support keys hosted on a ledger right now.
> I've wanted this feature for more than 1 year and I don't want to wait for SIGN_MODE_TEXTUAL and ledger implementation of the above.

alain2sf commented:
Just my 2 Cents... maybe I did not entirely grabbed and comprehend the use cases that @zmanian was describing above, so if my suggestion is not relevant just drop it ;) 

So here it is, would it be interesting to dig a bit on SSI? Like we could use a DID (dedicated/registered for Cosmos) and make pair-wise authN (holder <-> issuer), then we could use the Verifiable Credentials model, which is a Verifiable Claim kind of case where a digital signature (multi signature schemes are possible as well) is involved to prove anything (ownership or else). 

SSI: Self Sovereign Identity
DID: Decentralized Identifiers

zmanian commented:
The easiest use case is a forum or web application for delegators to a validator. If we had an standardized solution to this, I can as iqlusion's delegators to sign their logins and grant them access to the site.

It's definitely a thing that multiple validators have expressed interest in. It's not surprising to me that people outside the ecosystem don't need it.

These kinds of system are used extensively in the Etherum ecosystem now for linking token holding to offchain decision making processes.

ethanfrey commented:
@zmanian thank you for the use cases.

I think those would be fulfilled by the oauth-like proposal I made. Maybe that is overly complex and could be simplified or clarified.  But it had fields for site to log into, purpose of grant, and replay protection.

aaronc commented:
It may or may not be relevant but I wanted to share this _very_ simple module we have for Regen that I mentioned on the architecture call: https://github.com/regen-network/regen-ledger/blob/v0.6.0-alpha1/proto/regen/data/v1alpha1/tx.proto. It always for timestamping, signing, and possibly storing arbitrary data.

fdymylja commented:
On Friday, during community call, we were asked how this ADR would behave in regards to [ADR034](https://github.com/cosmos/cosmos-sdk/pull/7491). 

I've hit quite a roadblock here.

Since due to 034, it's not possible to assume anymore that `Address = Hash(PubKey)`

The process of verifying a message becomes:

`AccountPubKey = store.Get(msg.Signer).PubKey` (1)

This is fine for everything that needs to be verified on chain, but it's problematic to what needs to be verified off chain.

In particular, we could enrich the MsgSignedMessage with ChainID in order to get the `PubKey` as highlighted in (1). 
This is required because a signed message might have become invalid as `Signer` might have changed its `PubKey`

But `ChainID` does not offer information on what is the node IP to query in order to get `Account` information.

So whose duty is to map `ChainID` to `node's ip` to query? Is it the signature's verifiers job to do so?

If it's not their duty, who takes care of mapping `ChainID` to queriable nodes IPs? (This would introduce the need of a third party trust, which I think it's something we do not want)

alexanderbez commented:
Great question. Yes, the verifier would need to have a (trusted) list of nodes to query against for a given chain ID or it could listen for txs with messages that contain that message type and update it's local DB but this requires verifiers to be stateful.

ethanfrey commented:
I think that is a major issue that ADR 034 should address.

If they break this use case (which is not just for ADR 036, but for many client side code), they should provide a secure way of handling it . It may make all layer 2 solutions broken (imagine a state channel, but the pubkey changed in the middle - what key is needed to sign those packets, the old one or the new one).

Changing pubeys is easy to implement but brings in lots of problems to other components, which is why it was never done before. (I have not been following that discussion, maybe this has all been addressed, but if @fdymylja is hitting this issue here, I am sure plenty will hit it in other places). I kind of have the feeling ADR 034 breaks more than it fixes, and that subkeys/groups would solve most of the problems in a cleaner and more extensible way.

robert-zaremba commented:
Please see my comment from ADR-34. Without defending ADR-34, I want to highlight that it's important to design (if possible) for an abstraction above public key: https://github.com/cosmos/cosmos-sdk/pull/7491#issuecomment-724271462
I didn't read through the comments here, I will try to do it tomorrow and think about potential solutions.

ethanfrey commented:
> I want to highlight that it's important to design (if possible) for an abstraction beyond public key

I am not sure it is important. It is an interesting goal, and it requires lots of research to do it properly. But I see no strong need for it.

If there are little-to-no cases of other blockchains doing this, maybe we should take the time to reflect long and hard before trying to do so. I am happy for the discussion, I just think this is a larger research project to make a proper design.

robert-zaremba commented:
> If there are little-to-no cases of other blockchains doing this,  [...]

At least the following new blockchains implement account abstraction: Flow, Near, Algorand.... I see it as a value proposition of new blockchains. As stated in other thread, it provides clear separation between authorization and an ID. 
That being said I :100:  agree that we need to think and design it carefully. Thanks for comments @ethanfrey .

alessio commented:
Hi,

Just to make it absolutely clear: we intend to get on with this and at
least deliver on a subset of specifications/use cases with a first
implementation that is designed to address exactly those.

This is because we want to deliver a testable something that would allow us
to close the feedback loop as early as possible. That would also allows us
to have further meaningful conversations based on a concrete result.

Frojdi is on this and will execute accordingly.

Kindest regards.

   AT


On Mon, 9 Nov 2020, 22:44 Robert Zaremba, <notifications@github.com> wrote:

> If there are little-to-no cases of other blockchains doing this, [...]
>
> At least the following new blockchains implement account abstraction:
> Flow, Near, Algorand.... I see it as a value proposition of new
> blockchains. As stated in other thread, it provides clear separation
> between authorization and an ID.
> That being said I 💯 agree that we need to think and design it carefully.
> Thanks for comments @ethanfrey <https://github.com/ethanfrey> .
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-724322829>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABX73ED4KSIC7AEMFZRXCTSPBWDBANCNFSM4TDSDM7A>
> .
>

robert-zaremba commented:
@alessio  - what's the plan of shipping it and testing it? Where is it going to be deployed? A Gaia testnet? Cutoms chain?

robert-zaremba commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+
+## Abstract
+
+Currently, in the SDK, there is no convention to sign arbitrary message like on Ethereum. We propose with this specification a way to sign arbitrary message for Cosmos SDK chains.
```
```suggestion
Currently, in the SDK, there is no convention to sign arbitrary message like on Ethereum. We propose with this specification, for Cosmos SDK ecosystem, a way to sign and validate off-chain arbitrary messages.
```

robert-zaremba commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+
+The aim is being able to sign arbitrary messages, even using Ledger or similar HSM devices.
+
+As a result signed messages should look roughly like Cosmos SDK messages. chain-id, account_number and sequence can all be assigned invalid values.
```
```suggestion
As a result signed messages should look roughly like Cosmos SDK messages but **must not** be a valid on-chain transaction. `chain-id`, `account_number` and `sequence` can all be assigned invalid values.
```

robert-zaremba commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+
+Cosmos SDK 0.40 also introduces a concept of “auth_info” this can specify SIGN_MODES.
+
+A spec should include an auth info that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.
```
```suggestion
A spec should include an `auth_info` that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.
```

robert-zaremba commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+- chain-id should be equal to “signature”
+- fee gas should be equal to 0
+- fee amount should be an empty array
+- Inside the message with the type MsgSignText, we put inside a  *bytes* message and the address of the signer.
```
let's use MUST and MUST NOT in the spec above where possible.

Also, how about rewordgin the lines 33-42 and putting `auth_info` into the enumerated list, to have clear bullet points of requirements.

fdymylja replied:
I will address the changes

robert-zaremba commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,103 @@
+    bytes Signer = 1 [(gogoproto.jsontag) = "signer", (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
+    // Message represents the raw bytes of the content that is signed (text, json, etc)
+    bytes Message = 2 [(gogoproto.jsontag) = "message"];
+}
```
How about adding the a `TTL` (time to live) field? Or do we want to have a minimum structure of `bytes Message` that all messages will comply?

fdymylja replied:
The general idea was this. This was studied to serve as a general purpose message that can basically fit in anything. The reasons being: a cosmos message has around it all the tooling required for signing and verification, and also arbitrary data can fit in itself every use case, for instance `Message` could be any type of structured data, a text, a file.

robert-zaremba replied:
I can't imagine a use-case where we don't want TTL. But fair enough.
How about adding a discussion section to enumerate fields we think are necessary?

fdymylja replied:
A certificate which itself contains a TTL, plus there are different ways in which non replayability is achieved in systems, time isn't the only one. There are also nonce's systems etc.

robert-zaremba replied:
OK, fair enough for now.

robert-zaremba reviewed (changes requested):
- [ ] Describe how the verification works.
- [x] rename `MsgSignedMessage` or fix other name where we talk about it (personally I would drop Message for Data, as we use message for the `message` type).
- [x] Let's consider adding a TTL to `message MsgSignedMessage`
- [x] In some places, please reword from should('nt) to MUST or MUST NOT (example: `chain_id=0` makes sure that it won't be a valid on-chain transaction).
- [x] Please include more use-cases in the context based on the comments (https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-721062923, https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-722478477)
- [x] I would like to see an argument why we don't specify the codec for the message types. Leaving it open will create another 15 standards. We already use protobuf and JSON. I suggest to use JSON to let user see what he is signing. (BTW: while this ADR is in proposed state, we will be able t change the coding, but let's at least mention it).
- [ ] Suggestion: let's create a new section "further discussion" to list items suggested by @ethanfrey (the minimum data required to protect from replay attacks and token abuse).

alessio commented:
@robert-zaremba dixit:

> @alessio - what's the plan of shipping it and testing it?

As per Cosmos SDK standards, unit and integration testing plus client apps testing carried out in collaboration with partners and app developers.

> Where is it going to be deployed? A Gaia testnet? Cutoms chain?

The PR will be open against this repository's `master` branch first - as per Cosmos SDK's standard process. A Gaia testnet is an interesting idea, we're happy to do that. We would also be happy to backport the implementation to Cosmos SDK v0.39 release series, if there was demand for it.

robert-zaremba commented:
So, what would be a good idea, is to merge this PR as `Proposed`, make the release on testnet, and once we have feedback (as @alessio  suggested) go back, revisit and finalize this ADR. What do you think?

robert-zaremba commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,108 @@
+A spec should include an `auth_info` that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.
+
+- the memo must be empty
+- nonce, sequence number must be equal to 0
```
```suggestion
- nonce, sequence number must be equal to 0 (this make the transaction invalid on-chain)
```

robert-zaremba commented:
How about this suggestion:

>  I would like to see an argument why we don't specify the codec for the message types. Leaving it open will create another 15 standards. We already use protobuf and JSON. I suggest to use JSON to let user see what he is signing.

alessio commented:
@robert-zaremba dixit:

> So, what would be a good idea, is to merge this PR as Proposed, make the release on testnet, and once we have feedback (as @alessio suggested) go back, revisit and finalize this ADR

ACK from me. Sounds like we have a plan at last - thanks!

robert-zaremba commented:
@ethanfrey , @sunnya97 , @fdymylja  Re support for re-keying. Here is are 2 ideas how we can solve it:

1. We need to update the `MsgSignData` type with
    + block_height or block_id
    + merkle proof that a given key is assigned to given account at given block_height

   This complicate the verification

2. Update the ADR-34: each account which supports rekeying will have to remember old keys in the state.

fdymylja commented:
@robert-zaremba 
> I would like to see an argument why we don't specify the codec for the message types. Leaving it open will create another 15 standards. We already use protobuf and JSON. I suggest to use JSON to let user see what he is signing.

AFAIK the encoding of the message does not matter, if it goes through a gRPC verification endpoint it can be decoded via protobuf unmarshalling, if it goes through a rest verification endpoint it can be verified via JSON unmarshalling. If I'm not mistaken `TxBytes` are decoded into `sdk.Tx` and then the `SignatureBytes` are produced from `sdk.Tx`. So the message encoding should not matter.

As per the `Data`, even in this case I think it should be developer's choice. If an off-chain signature is targeted at users, and should be signed by users who have to ACK what they're signing the `Data` bytes can be stringified JSON. 

On the contrary if the verifier is software, which needs to unmarshal bytes into an `object`, it can do it choosing the serializing solution it prefers.

> I can't imagine a use-case where we don't want TTL. But fair enough.

Just to note this here too, `time to live` based security is not the only used in non-replayable security systems. For example nonces are a valid alternative too.

fdymylja commented:
@robert-zaremba 
> @ethanfrey , @sunnya97 , @fdymylja Re support for re-keying. Here is are 2 ideas how we can solve it:
> 
> 1. We need to update the `MsgSignData` type with
>    
>    * block_height or block_id
>    * merkle proof that a given key is assigned to given account at given block_height
>    
>    This complicate the verification
> 2. Update the ADR-34: each account which supports rekeying will have to remember old keys in the state.

I still think this change would break current offchain implementations. 

Regarding future implementations, the biggest issues I see are the following:

1. You need to map chain IDs too, as given address can have two different pubkeys based on the chain it is pointing to.
2. You need to map chain IDs to queriable nodes.

Number 2 can be solved by being stateful, hence listening and keeping a ledger of pub key changes related to addresses, as @alexanderbez  suggested.

robert-zaremba commented:
> As per the `Data`, even in this case I think it should be developer's choice. If an off-chain signature is targeted at users, and should be signed by users who have to ACK what they're signing the `Data` bytes can be stringified JSON.

Yes, I meant about `Data` (in the text originally it was message, hence I used that in the previous comment).
The motivation to use a single serialization for data content is to standarize the tools. The the software can check whatever dynamic fields he likes to authorize the user.

> On the contrary if the verifier is software, which needs to unmarshal bytes into an object, it can do it choosing the serializing solution it prefers.

So this is the point - a verifier can't choose the serialization it prefers. It must use exactly same serialization as the message author.

fdymylja commented:
@robert-zaremba 
> So this is the point - a verifier can't choose the serialization it prefers. It must use exactly same serialization as the message author.

My assumption is the contrary. 

The prover must abide by the rules the verifier imposes on `Data`.

This assumption might be problematic only in the case for which the verifier wants to enable multiple encodings of `Data` passing through the same verification endpoint. 

As the `content-type` policy should be meaningful only for how `MsgSignedData` must be decoded, leaving `Data` decoding mechanism unknown.

But I think this use case would be really an edge case.

robert-zaremba commented:
Shall we move this PR and merge it as discussed [here](https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-724804178)?

From my side, this [checklist](https://github.com/cosmos/cosmos-sdk/pull/7727#pullrequestreview-527096649) is still open.

fdymylja commented:
@robert-zaremba 
> describe how verification works

I didn't understand this part, I've create a small example of ADR036 implementation which exposes two rest api endpoints one for verification and one for signing, using the tentative implementation. You can have a look here, there's also the postman collection: https://github.com/fdymylja/adr036-example/blob/master/main.go

robert-zaremba commented:
>> describe how verification works
>
> I didn't understand this part,

@fdymylja  - please describe in this ADR how verification should work.

fdymylja commented:
@robert-zaremba 
> @fdymylja - please describe in this ADR how verification should work.

I've added the description on how verification works

alessio commented:
@alexanderbez you requested changes, can you review this again please?

robert-zaremba commented:
@fdymylja I was expecting more details in the verification description then simple _follows the same rules as an onchain one_.
On chain verification is doing other operations. Here, at minimum we need:
+ adding a prefix to the message signing to avoid malleability
+ Do we validate against account address or account key? What are we going to use for query?

robert-zaremba commented:
If this is work in progress, then you can probably add TODO and we can try to sort it out later. Thoughts?

alessio commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,112 @@
+
+### Negative
+
+
```
```suggestion
## Further discussion
```

alessio commented:
> let's create a new section "further discussion"

This is now created. Do you want to add something to it @ethanfrey @robert-zaremba?

robert-zaremba commented:
> > let's create a new section "further discussion"
> 
> This is now created. Do you want to add something to it @ethanfrey @robert-zaremba? If so, please feel free to add directly on the file +1

My suggestion was to:
> let's create a new section "further discussion" to list items suggested by @ethanfrey (the minimum data required to protect from replay attacks and token abuse).

rather than having an empty section. Just a simple list of problems / issues and solutions (solutions can come in the future) to elaborate  later on.

robert-zaremba commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,112 @@
+- It gives room for other off-chain messages specifications that aim to target more specific and common use cases such as off-chain-based authN/authZ layers [2].
+
+### Negative
+
```
```suggestion

+ Current proposal requires a fixed relation between account address and a public key.
+ Doesn't work with multi sig accounts.
```

robert-zaremba replied:
Please feel free to add more negative consequences - we were discussing it in the thread. Leaving it empty is not good.

alessio commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,114 @@
+
++ Current proposal requires a fixed relationship between an account address and a public key.
++ Doesn't work with multi sig accounts.
+## Further discussion
```
```suggestion

## Further discussion
```

alessio commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,114 @@
+
+### Negative
+
++ Current proposal requires a fixed relationship between an account address and a public key.
```
```suggestion
- Current proposal requires a fixed relationship between an account address and a public key.
```

alessio commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,114 @@
+### Negative
+
++ Current proposal requires a fixed relationship between an account address and a public key.
++ Doesn't work with multi sig accounts.
```
```suggestion
- Doesn't work with multi sig accounts.

```

alessio commented on docs/architecture/adr-036-arbitrary-signature.md:
```diff
@@ -0,0 +1,115 @@
+### Negative
+
+- Current proposal requires a fixed relationship between an account address and a public key.
+- Doesn't work with multi sig accounts.
```
```suggestion
- Doesn't work with multisig accounts.
```

alessio commented:
> Please include more use-cases in the context based on the comments (#7727 (comment), #7727 (comment))

@robert-zaremba use cases are already mentioned with URLs in the PR.

robert-zaremba reviewed (approved):
Approving, but the analysis in this PR should have been better.

alessio commented:
> but the analysis in this PR should have been better.

This will certainly teach @zmanian, @antoineherzog and the other people that worked on this a lesson

moshemalawach commented:
now that's rude. is this how you encourage opensource collaboration?



\-------- Message d'origine --------
Le 13 nov. 2020 à 18:28, Alessio Treglia < notifications@github.com > a écrit :

>
>
>
> > but the analysis in this PR should have been better.
>
> This will certainly teach [@zmanian][zmanian], [@antoineherzog][antoineherzog] and the other people that worked on this a lesson
>
> —
> You are receiving this because you commented.
> Reply to this email directly, [view it on GitHub][], or [unsubscribe][].![AJXOYW5PM6PJAFIKOUDJWGTSPVUDFA5CNFSM4TDSDM7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFNJY3LI.gif][]


[zmanian]: https://github.com/zmanian
[antoineherzog]: https://github.com/antoineherzog
[view it on GitHub]: https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-726896045
[unsubscribe]: https://github.com/notifications/unsubscribe-auth/AJXOYW7QJJHL6RE2IR6Q3LDSPVUDFANCNFSM4TDSDM7A
[AJXOYW5PM6PJAFIKOUDJWGTSPVUDFA5CNFSM4TDSDM7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFNJY3LI.gif]: https://github.com/notifications/beacon/AJXOYW5PM6PJAFIKOUDJWGTSPVUDFA5CNFSM4TDSDM7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFNJY3LI.gif

alessio commented:
> now that's rude.

The expression `that's rude` couldn't not come from somebody that totally
lacks knowledge of the context or the relationships between the parties
involved (Antoine, Zaki and myself).

> is this how you encourage opensource collaboration?

I was being sarcastic and you seem to be looking for a fight. It's
understandable in these challenging times where most of us are locked in in
our homes. But I won't bite ;)

On Fri, Nov 13, 2020 at 7:25 PM Moshe Malawach <notifications@github.com>
wrote:

> now that's rude. is this how you encourage opensource collaboration?
>
>
>
> \-------- Message d'origine --------
> Le 13 nov. 2020 à 18:28, Alessio Treglia < notifications@github.com > a
> écrit :
>
> >
> >
> >
> > > but the analysis in this PR should have been better.
> >
> > This will certainly teach [@zmanian][zmanian],
> [@antoineherzog][antoineherzog] and the other people that worked on this a
> lesson
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, [view it on GitHub][], or
> [unsubscribe][].![AJXOYW5PM6PJAFIKOUDJWGTSPVUDFA5CNFSM4TDSDM7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFNJY3LI.gif][]
>
>
> [zmanian]: https://github.com/zmanian
> [antoineherzog]: https://github.com/antoineherzog
> [view it on GitHub]:
> https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-726896045
> [unsubscribe]:
> https://github.com/notifications/unsubscribe-auth/AJXOYW7QJJHL6RE2IR6Q3LDSPVUDFANCNFSM4TDSDM7A
> [AJXOYW5PM6PJAFIKOUDJWGTSPVUDFA5CNFSM4TDSDM7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFNJY3LI.gif]:
>
> https://github.com/notifications/beacon/AJXOYW5PM6PJAFIKOUDJWGTSPVUDFA5CNFSM4TDSDM7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFNJY3LI.gif
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-726988131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABX73HDJE7Z74T3CAI6ZK3SPWB3VANCNFSM4TDSDM7A>
> .
>


-- 

*Alessio Treglia*

Director of Engineering at Tendermint <https://tendermint.com/>
Cosmonaut at Cosmos <https://cosmos.network/>

[image: ✨] @tendermint_team <https://twitter.com/tendermint_team> · @cosmos
<http://twitter.com/cosmos> · *We're hiring!*
<https://tendermint.com/careers/> [image: 🚀]

zmanian commented:
One thing I think we should add here is that for verification purposes the account number and the sequence are always `0` in verification for the purposes of this adr.

alessio commented:
> One thing I think we should add here is that for verification purposes the account number and the sequence are always 0 in verification for the purposes of this adr.

Makes sense. We'll add it :+1:

moshemalawach commented:
> you seem to be looking for a fight

Definitely not, and this is why I immediately removed my message on a second thought after posting it as it wasn't the goal. But looks like it arrived by email anyway :)

alessio commented:
@moshemalawach That's fine, now I know. Peace! :pray:

marbar3778 commented:
Progress? Please close the pr if it's not ready to be merged

alessio commented:
@antoineherzog I am afraid we can't continue doing this work because there's ostensibly a lack of priorities alignment across the various teams that work on the Cosmos SDK.

That said @fdymylja, great effort. Keep up the great work!

marbar3778 commented:
> I am afraid we can't continue doing this work because there's ostensibly a lack of priorities alignment across the various teams that work on the Cosmos SDK.

I would say this statement is false. **No one** has taken the initiative to move this forward.

Anyways, we are here to collaborate. Let's figure out a way we can move forward with this. We can include the discussion points in an upcoming sdk architecture call. And coordinate a call with the necessary parties if they are unable to attend the sdk calls.

zmanian commented:
So I think really the only open issue is how to handle stateful validation criteria that might be introduced into the SDK by parallel contemporaneous work.

Here is what I would propose.

The Verifier should provide an optional validation context to the signer.
```
"validation_context":"chain-id", "height"
```

The signer should sign over the validation context and provide correct credentials.

The verifier will need to check the signed validation context is as expected and then pull the account state from the chain-id and height to verify.

zmanian commented:
Sommelier has a near term usecase for this authorization mechanism.

If there isn't support for it, we will do it out of tree

alessio commented:
> > I am afraid we can't continue doing this work because there's ostensibly a lack of priorities alignment across the various teams that work on the Cosmos SDK.
> 
> I would say this statement is false. **No one** has taken the initiative to move this forward.
> 

@fdymylja has. He opened a PR that implemented this ADR.

I want to add that unlike other ADRs, whose implementations were merged even before their respective ADRs were approved, the implementation PR got blocked because the ADR committee failed to reach a consensus.

I'm deliberately not pointing my finger at anyone. I believe that no individual developers are at fault here. This ADR is just the brightest example of the damage that the utter lack of clear governance can do to an ever-growing and complex open source project like Cosmos SDK.

_My 2p_

alessio commented:
> If there isn't support for it, we will do it out of tree

I would welcome such an initiative. Please do reach out, we're very happy to collaborate on an out-of-tree solution.

moshemalawach commented:
> Sommelier has a near term usecase for this authorization mechanism.
> 
> If there isn't support for it, we will do it out of tree

At aleph.im we already using it in production according to what has been implemented in first proposal.

robert-zaremba commented:
> I want to add that unlike other ADRs, whose implementations were merged even before their respective ADRs had been approved, the implementation PR got blocked because the ADR committee failed to reach a consensus.

Is there still anything blocking this ADR? If yest, let's add it to the Friday Architecture Call agenda. This is the easiest way to unblock ADRs.

marbar3778 commented:
We have now reached consensus. We have the three approvals needed and then some. Let's merge this PR.

aaronc commented:
> I want to add that unlike other ADRs, whose implementations were merged even before their respective ADRs had been approved, the implementation PR got blocked because the ADR committee failed to reach a consensus.
> 
> I'm deliberately not pointing my finger at anyone. I believe that no individual developers are at fault here. This ADR is just the brightest example of the damage that the utter lack of clear governance can do to an ever-growing and complex open source project like Cosmos SDK.

Wow, I'm sensing you and possibly your team are pretty frustrated by the process here, and I'm sorry to hear that. This is a complex project in complex times and I definitely appreciate the challenges in working on a multi-entity, multi-stakeholder open source project. I also apologize for any way that I or my team may have contributed to that frustration, either directly or indirectly.

That said, I feel it is important to note that as far as I can tell @marbar3778's original observation that no one picked up this work seems pretty accurate. From reviewing the commit log and also the meeting notes where this ADR was discussed, it seems pretty clear that nobody worked on this or #7896 after Nov 16th, 2020. I did find references in meeting notes from Nov 20th and Feb 26th indicating that your team intended to spend more time on this with the aim to reconcile it with ADR 034, but I have not seen any progress on that. I understand there are things in this development process that can be upsetting. It seems like the back and forth on Nov 13th got a little tense even though it ended alright. All of that aside, I do not think it is accurate or fair to say that the ADR committee blocked this or the implementation because of a failure to reach consensus. There _were_ and still are outstanding issues including compatibility with ADR 034, and now that this has been merged they _still will_ need to be addressed. If your team needed to move this forward quickly as sort of a "preview" feature, I and I believe the rest of my team would have been open to that. From my perspective, we had given the prelimary concept ACK both here and in architecture meetings and were waiting for your team to take the next steps.

Without a doubt we need improve the way we do things as a complex, multi-stakeholder project. At the same time we are all a part of the process. If you’re feeling blocked, if things aren’t going well in a discussion thread, or what not, please reach out or bring it up on a call. We’re here to collaborate and improve things and we want make this project work for all stakeholders.

alessio commented:
> Without a doubt we need improve the way we do things as a complex, multi-stakeholder project. At the same time we are all a part of the process. If you’re feeling blocked, if things aren’t going well in a discussion thread, or what not, please reach out or bring it up on a call. We’re here to collaborate and improve things and we want make this project work for all stakeholders.

I believe there are still governance issues here that have not yet been resolved (and neither of our respective teams are part of the problem here). I tried to contribute to the conversations and advised on the subject multiple times. Nothing happened afterward.

That said, I appreciate your response wholeheartedly. We remain committed to supporting the Cosmos SDK development on a best-effort basis.

Thanks, and keep up the great work!

webmaster128 commented:
Nice work. Just to double check: do we agree that multiple messages are supported by this spec? This is somewhat implied by not having a message restriction in the dummy tx.

robert-zaremba commented:
@webmaster128 this is not for signing transaction - we don't want to do it here. This is for signing messages which are not transactions.

webmaster128 commented:
@robert-zaremba sure, but if you wrap a special message type in an `msg` array of a (fake) `cosmos-sdk/StdTx`, we should either support multiple messages or add an explicit restriction that `msg` must contain exactly one element in order to be valid.

webmaster128 commented:
In https://github.com/cosmos/cosmjs/pull/847 you find is a CosmJS implementation of this spec. Pretty straight forward if Amino signing tooling is present.

The main open question I have is how to encode the signed data, i.e. the **exchange format** from signer to verifier. I see the followiging options:
1. The pair (`data`, `raw_secp256k1_signature`). This is how Ethereuum's web3.eth.sign works.
2. The pair (`data`, `signature` as StdSignature). Like 1. but is open for Non-Secp256k1 signature algorithms.
3. The whole StdTx structure as shown in this ADR with top level `"type": "cosmos-sdk/StdTx", "value": …`.
4. The content of StdTx's `value` field. This is currently used in my implementation.

Option 3 and 4 allow using multiple datas through multiple messages (see https://github.com/cosmos/cosmjs/pull/847/commits/06da6b8ed79fa107fc49ece029c7f701f7eb9f8a). I'm not saying this is useful but seems to work without violating the spec.

fdymylja commented:
> In [cosmos/cosmjs#847](https://github.com/cosmos/cosmjs/pull/847) you find is a CosmJS implementation of this spec. Pretty straight forward if Amino signing tooling is present.
> 
> The main open question I have is how to encode the signed data, i.e. the **exchange format** from signer to verifier. I see the followiging options:
> 
> 1. The pair (`data`, `raw_secp256k1_signature`). This is how Ethereuum's web3.eth.sign works.
> 2. The pair (`data`, `signature` as StdSignature). Like 1. but is open for Non-Secp256k1 signature algorithms.
> 3. The whole StdTx structure as shown in this ADR with top level `"type": "cosmos-sdk/StdTx", "value": …`.
> 4. The content of StdTx's `value` field. This is currently used in my implementation.
> 
> Option 3 and 4 allow using multiple datas through multiple messages (see [cosmos/cosmjs@06da6b8](https://github.com/cosmos/cosmjs/commit/06da6b8ed79fa107fc49ece029c7f701f7eb9f8a)). I'm not saying this is useful but seems to work without violating the spec.

This spec is not updated, and needs to be updated, we've had a call weeks ago, and we decide to generalize the purpose of this ADR further.

In the sense that ADR-036 should not aim to provide a predefined Message such as `MsgSignData`, the only aim for this ADR is to provide a spec that can be understood and used by existing tooling (cosmjs, ledger etc), such that an offchain transaction (which is a composition of messages) looks exactly like a normal cosmos-sdk TX with the only property of being always invalid if run on a live chain.

This is the reference to the call we've had: https://github.com/cosmos/cosmos-sdk/pull/7896#issuecomment-847886405

assafmo commented:
For reference, this is what we ended up doing on Secret Network:
- Implementation: https://github.com/enigmampc/cosmos-sdk/compare/v0.9.1-scrt...v0.9.2-scrt
- Keplr & CLI usage examples: https://github.com/enigmampc/snip20-reference-impl/pull/22
